### PR TITLE
domain: Close DDL when creating a domain failed

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -453,11 +452,20 @@ func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duratio
 	}
 	sysCtxPool := pools.NewResourcePool(sysFac, 2, 2, idleTimeout)
 	d.ddl = ddl.NewDDL(ctx, d.etcdClient, d.store, d.infoHandle, callback, ddlLease, sysCtxPool)
+	var err error
+	defer func() {
+		if err != nil {
+			d.Close()
```
Comment (shenli) [inline domain/domain.go]: Add comment here.

**Comment by shenli** [comment]
@coocood @winkyao @breeswish PTAL

**Comment by winkyao** [comment]
@zimulala add some pr description?

File: domain/domain.go [hunk 2]
```diff
@@ -409,10 +409,10 @@ type EtcdBackend interface {
 }
 
 // NewDomain creates a new domain. Should not create multiple domains for the same store.
-func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duration, factory pools.Factory, sysFactory func(*Domain) (pools.Resource, error)) (d *Domain, err error) {
+func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duration, factory pools.Factory, sysFactory func(*Domain) (pools.Resource, error)) (*Domain, error) {
```
Comment (tiancaiamao) [inline domain/domain.go]: I think this code is a bit of chaos, which lead to the source of bugs.
Would it be better if we split `NewDomain` into two parts: create object, and then initialize?
`NewDomain()` will return the object, with no error.
`domain.Init()` will initialize all kind of things, if it return error, we **must** close the domain object.

**Comment by shenli** [comment]
@tiancaiamao We should fix this issue quickly and then try to improve this. Please discuss with @zimulala and file a issue.

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
@shenli @tiancaiamao 
Done.

**Pull request merged**